File: src/test/scala/org/egonet/model/question/QuestionTest.scala
package org.egonet.model.question


import org.scalatest.Args
import org.egonet.EgonetSpec
import org.egonet.exceptions.MalformedQuestionException
import org.egonet.model.Shared
import org.egonet.model.answer.Answer
import org.egonet.model.answer.CategoricalAnswer

class QuestionTest extends EgonetSpec {

  "Question utility methods" should "correctly instantiate question types" in {
    
    intercept[MalformedQuestionException] {
       Question.asSubclass("please-dont-exist-as-a-class")
    }
    
    for(clazz <- Shared.questionClasses) {
    
      var clazzCanonicalName = clazz.getCanonicalName()
      
      assert(Question.newInstance(clazz).getClass().equals(clazz), 
          "Question.newInstance(" + clazzCanonicalName + ") should yield an instance of that type" )
    }
  }
    
    "Question getters and setters" should "set and get the same values" in {
      
      assert(new StudyQuestion().getTitle() == "Study questions")     
      assert(new EgoQuestion().getTitle() == "Questions About You")
      assert(new AlterQuestion().getTitle() == "<html><p>Questions About <nobr><b>$$1</b></nobr></p></html>")
      assert(new AlterPromptQuestion().getTitle() == "Whom do you know?")
      assert(new AlterPairQuestion().getTitle() == "<html><p>Questions About <nobr><b>$$1</b></nobr> and <nobr><b>$$2</b></nobr></p></html>")
      
      val q = new StudyQuestion("ignore")
    	List(true,false).foreach(t => { 
    	  q.setFollowupOnly(t)
    	  assert(q.isFollowupOnly == t)
    	})
    	
    	List(true,false).foreach(t => { 
    	  q.setStatable(t)
    	  assert(q.isStatable == t)
    	})
    	
    	val a = new CategoricalAnswer()
    	q.setAnswer(a)
    	assert(q.getAnswer() == a)
    	
    	val sAdjacent = new Selection()
      	sAdjacent.setAdjacent(true)
      	sAdjacent.setValue(12345)
      
      	 import scala.collection.JavaConverters._
    	val selections = List(
    	    new Selection(),
    	    sAdjacent, // determinesAdjacency=true
    	    new Selection()
    	    )
    	q.setSelections(selections.asJava)
    	
    	val rSelections = q.getSelections()
    	val zipped = selections.toArray.zipWithIndex
    	zipped.foreach({case (x,i) => {
    	  assert(x == rSelections.get(i))
    	}})
    	
    	assert(q.determinesAdjacency())
    	assert(q.selectionAdjacent(12345))
    	
    }
    
    
  
}

File: src/test/java/org/egonet/model/question/QuestionTest.java
package org.egonet.model.question;

import org.egonet.exceptions.MalformedQuestionException;
import org.egonet.io.StudyReader;
import org.egonet.model.Answer;
import org.egonet.model.Question;
import org.egonet.model.Selection;
import org.egonet.model.Shared;
import org.egonet.model.Shared.QuestionType;
import org.junit.Test;

import static org.junit.Assert.*;

import java.util.Arrays;
import junit.framework.JUnit4TestAdapter;

public class QuestionTest {
	public static junit.framework.Test suite() {
	      return new JUnit4TestAdapter(QuestionTest.class);
	}

	@Test(expected=MalformedQuestionException.class)
	public void testBadQuestionClass() {
		Question.asSubclass("please-dont-exist-as-a-class");
	}

	@Test
	public void testQuestionTypesMatch() {
	    for(QuestionType type : Shared.QuestionType.values()) {

	        String clazzCanonicalName = type.name();

	        assertTrue("Question.newInstance(" + clazzCanonicalName + ") should yield an instance of that type",
	        		StudyReader.questionTypeFromString(clazzCanonicalName).equals(type));
	    }
	}

	@Test
	public void testQuestionTitles() {
	      assertTrue(new Question(QuestionType.STUDY_CONFIG).getTitle() == "Study questions");
	      assertTrue(new Question(QuestionType.EGO).getTitle() == "Questions About You");
	      assertTrue(new Question(QuestionType.ALTER).getTitle() == "<html><p>Questions About <nobr><b>$$1</b></nobr></p></html>");
	      assertTrue(new Question(QuestionType.ALTER_PROMPT).getTitle() == "Whom do you know?");
	      assertTrue(new Question(QuestionType.ALTER_PAIR).getTitle() == "<html><p>Questions About <nobr><b>$$1</b></nobr> and <nobr><b>$$2</b></nobr></p></html>");
	}

	@Test
	public void testGettersSetters() {
		Question q = new Question(QuestionType.STUDY_CONFIG);
    	for(boolean t : new boolean[]{true,false}) {
    	  q.setFollowupOnly(t);
      	  assertTrue(q.isFollowupOnly() == t);
    	}

    	for(boolean t : new boolean[]{true,false}) {
    	  q.setStatable(t);
    	  assertTrue(q.isStatable() == t);
    	}

    	Answer a = new Answer();
    	q.setAnswer(a);
    	assertTrue(q.getAnswer() == a);

    	Selection sAdjacent = new Selection();
      	sAdjacent.setAdjacent(true);
      	sAdjacent.setValue(12345);

      	Selection [] selections = new Selection[]{
    	    new Selection(),
    	    sAdjacent, // determinesAdjacency=true
    	    new Selection()
    	};
    	q.setSelections(Arrays.asList(selections));

    	Selection [] rSelections = q.getSelections().toArray(new Selection[0]);
    	for(int i = 0; i < selections.length; i++) {
    	  assertTrue(selections[i] == rSelections[i]);
    	}

    	assert(q.determinesAdjacency());
    	assert(q.selectionAdjacent(12345));
	}
}
